File: packages/core/servers/discovery.ts
/** mDNS service discovery for LocalDub servers.

Server types:
  - `torch`              → pytorch_server.py (ASR + separate)  → default port 19109
  - `voxcpm_torch_gradio` → voxcpm_torch_server/server.py (TTS) → default port 19112

Usage:
  ```ts
  import { findServer, findServers } from '@repo/core/servers/discovery'

  // Get the first discovered (or fallback) server
  const { port, host } = await findServer('torch', 19109)

  // Get all discovered servers as base URLs
  const urls = await findServers('torch', 19109)
  ```
*/
import type { ServerType } from '@repo/config/servers'
import { SERVICE_MAP } from '@repo/config/servers'

export interface ServerInfo {
  host: string
  port: number
  foundVia: 'mdns' | 'default' | 'portfile'
}


/** Timeout for mDNS browse (ms). */
const MDNS_TIMEOUT = 3000


/**
 * Discover all running instances of a server type via mDNS.
 * Falls back to a single default entry if nothing found.
 *
 * Returns base URLs like `http://127.0.0.1:19109`.
 */
export async function findServers(
  type: ServerType,
): Promise<string[]> {
  const mdnsList = await findServerViaMdnsAll(type, MDNS_TIMEOUT)
  if (mdnsList.length > 0) {
    return mdnsList.map((s) => `http://${s.host}:${s.port}`)
  }
  return [`http://127.0.0.1:${type === 'voxcpm_torch_gradio' ? 19112 : 19109}`]
}

/**
 * Discover a single LocalDub server by mDNS, falling back to default port.
 */
export async function findServer(
  type: ServerType = 'voxcpm_torch_gradio',
): Promise<ServerInfo> {
  console.log(`findServer(${type})`)
  const mdnsList = await findServerViaMdnsAll(type, MDNS_TIMEOUT)
  console.log(`findServer(${type}) => mdnsList=${JSON.stringify(mdnsList)}`)
  if (mdnsList.length > 0) {
    return { ...mdnsList[0], foundVia: 'mdns' }
  }
  return { host: '127.0.0.1', port: type === 'voxcpm_torch_gradio' ? 19112 : 19109, foundVia: 'default' }
}

export async function findServerViaMdnsAll(
  type: ServerType,
  timeoutMs: number=MDNS_TIMEOUT,
): Promise<{ host: string; port: number }[]> {
  const serviceType = SERVICE_MAP[type]
  if (!serviceType) return []

  const Bonjour = await import('bonjour-service').then((m) => m.Bonjour)
  const bonjour = new Bonjour()
  const results: { host: string; port: number }[] = []

  return new Promise((resolve) => {
    const timer = setTimeout(() => {
      browser.stop()
      bonjour.destroy()
      resolve(results)
    }, timeoutMs)

    const browser = bonjour.find({ type: serviceType.replace(/^_|\._tcp\.local$/g, '') }, (svc) => {
      const entry = { host: svc.referer?.address ?? '127.0.0.1', port: svc.port }
      // Deduplicate
      if (!results.some((r) => r.host === entry.host && r.port === entry.port)) {
        results.push(entry)
      }
    })

    browser.start()
  })
}

/** Read the first PORT= line from a spawned process stdout. */
export function readPortFromOutput(output: string): number | null {
  const m = output.match(/^PORT=(\d+)/m)
  return m ? parseInt(m[1], 10) : null
}

File: packages/core-rs/src/servers/discovery.rs
use std::time::Duration;

use config_rs::servers::ServerType;
use mdns_sd::{ServiceDaemon, ServiceEvent};

const MDNS_TIMEOUT: Duration = Duration::from_millis(3000);

const DEFAULT_HOST: &str = "127.0.0.1";

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum FoundVia {
    Mdns,
    Default,
    PortFile,
}

#[derive(Debug, Clone)]
pub struct ServerInfo {
    pub host: String,
    pub port: u16,
    pub found_via: FoundVia,
}

/// Discover all matching servers via mDNS, falling back to default host/port.
///
/// Returns base URLs like `http://127.0.0.1:19109`.
pub async fn find_servers(type_: ServerType) -> Vec<String> {
    let list = find_server_via_mdns_all(type_, None).await;
    if !list.is_empty() {
        return list
            .into_iter()
            .map(|(host, port)| format!("http://{}:{}", host, port))
            .collect();
    }
    vec![format!("http://{}:{}", DEFAULT_HOST, type_.default_port())]
}

/// Discover a single server by mDNS, falling back to defaults.
pub async fn find_server(type_: ServerType) -> ServerInfo {
    let list = find_server_via_mdns_all(type_, None).await;
    if let Some((host, port)) = list.into_iter().next() {
        return ServerInfo {
            host,
            port,
            found_via: FoundVia::Mdns,
        };
    }
    ServerInfo {
        host: DEFAULT_HOST.to_string(),
        port: type_.default_port(),
        found_via: FoundVia::Default,
    }
}

/// Browse mDNS for the given server type.
///
/// Returns `(ip, port)` pairs for all resolved services of the matching type.
pub async fn find_server_via_mdns_all(
    type_: ServerType,
    timeout: Option<Duration>,
) -> Vec<(String, u16)> {
    let timeout = timeout.unwrap_or(MDNS_TIMEOUT);

    let Ok(daemon) = ServiceDaemon::new() else {
        return vec![];
    };
    let Ok(receiver) = daemon.browse(type_.service_name()) else {
        return vec![];
    };

    let start = std::time::Instant::now();
    let mut results: Vec<(String, u16)> = vec![];

    while start.elapsed() < timeout {
        let remaining = timeout.saturating_sub(start.elapsed());
        if remaining.is_zero() {
            break;
        }
        let recv_timeout = remaining.min(Duration::from_millis(200));
        match receiver.recv_timeout(recv_timeout) {
            Ok(ServiceEvent::ServiceResolved(info)) => {
                let port = info.get_port();
                for addr in info.get_addresses() {
                    let ip = addr.to_string();
                    let entry = (ip, port);
                    if !results.contains(&entry) {
                        results.push(entry);
                    }
                }
            }
            _ => {}
        }
    }

    drop(daemon);
    results
}

/// Read the first `PORT=XXXX` line from process stdout.
pub fn read_port_from_output(output: &str) -> Option<u16> {
    for line in output.lines() {
        if let Some(port_str) = line.strip_prefix("PORT=") {
            if let Ok(port) = port_str.trim().parse::<u16>() {
                return Some(port);
            }
        }
    }
    None
}
